File: powx-n.java
public class Solution {
    public double pow(double x, int n) {
        if (x == 0) return 0;
        if (n == 0) return 1;
        double ans = 1;
        if (n < 0) {
            if (n == Integer.MIN_VALUE) {
                n = n + 1;
                ans = 1 / x;
            }
            x = 1 / x;
            n = -n;
        }
        double power = x;
        while (n > 0) {
            if (n % 2 != 0) ans *= power;
            n >>= 1;
            power *= power;
        }
        return ans;
    }
}

File: powx-n.cc
class Solution {
public:
    double pow(double x, int n) {
        if (n == 0) return 1;
        if (x == 0) return 0;
        return compute(x, n);
    }
    
    double compute(double x, int n) {
        if (n == 1) return x;
        if (n == -1) return 1 / x;
        double ret = compute(x, n / 2);
        ret *= ret;
        if (n % 2) ret *= n > 0 ? x : 1 / x;
        return ret;
    }
};
